File: AndBible/test/net/bible/service/format/osistohtml/VerseHandlerTest.java
package net.bible.service.format.osistohtml;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;
import net.bible.service.format.osistohtml.OsisToHtmlSaxHandler.VerseInfo;

import org.crosswire.jsword.book.OSISUtil;
import org.junit.Before;
import org.junit.Test;
import org.xml.sax.helpers.AttributesImpl;

public class VerseHandlerTest {
	private OsisToHtmlParameters osisToHtmlParameters;
	private VerseInfo verseInfo;
	private BookmarkMarker bookmarkMarker;
	private MyNoteMarker myNoteMarker;
	private HtmlTextWriter htmlTextWriter;
	
	private VerseHandler verseHandler;
	
	@Before
	public void setUp() throws Exception {
		osisToHtmlParameters = new OsisToHtmlParameters();
		verseInfo = new VerseInfo();
		bookmarkMarker = new BookmarkMarker(osisToHtmlParameters, verseInfo, htmlTextWriter);
		myNoteMarker = new MyNoteMarker(osisToHtmlParameters, verseInfo, htmlTextWriter);
		htmlTextWriter = new HtmlTextWriter();
		
		verseHandler = new VerseHandler(osisToHtmlParameters, verseInfo, bookmarkMarker, myNoteMarker, htmlTextWriter);
	}

	/**
	 * No attributes. Just start verse, write some content, end verse.
	 * 
	 * <title>The creation</title>
	 */
	@Test
	public void testSimpleVerse() {
		osisToHtmlParameters.setShowVerseNumbers(true);
		AttributesImpl attrs = new AttributesImpl();
		attrs.addAttribute(null, null, OSISUtil.OSIS_ATTR_OSISID, null, "Ezek.40.5");	
		verseHandler.start(attrs);
		htmlTextWriter.write("The Creation");
		verseInfo.isTextSinceVerse = true;
		verseHandler.end();
		
		assertThat(htmlTextWriter.getHtml(), equalTo(" <span class='verse' id='5'><span class='verseNo'>5</span>&#160;The Creation</span>"));
	}
}

File: app/src/test/java/net/bible/service/format/osistohtml/taghandler/VerseHandlerTest.kt
package net.bible.service.format.osistohtml.taghandler

import net.bible.service.format.osistohtml.HtmlTextWriter
import net.bible.service.format.osistohtml.OsisToHtmlParameters
import net.bible.service.format.osistohtml.osishandlers.OsisToHtmlSaxHandler.VerseInfo
import org.crosswire.jsword.book.OSISUtil
import org.hamcrest.CoreMatchers
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.xml.sax.helpers.AttributesImpl
import java.util.*

class VerseHandlerTest {
    private var osisToHtmlParameters: OsisToHtmlParameters? = null
    private var verseInfo: VerseInfo? = null
    private var bookmarkMarkerMock: BookmarkMarker? = null
    private var myNoteMarker: MyNoteMarker? = null
    private var htmlTextWriter: HtmlTextWriter? = null
    private var verseHandler: VerseHandler? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        osisToHtmlParameters = OsisToHtmlParameters()
        osisToHtmlParameters!!.chapter = 3
        osisToHtmlParameters!!.isBible = true
        verseInfo = VerseInfo()
        bookmarkMarkerMock = Mockito.mock(BookmarkMarker::class.java)
        htmlTextWriter = HtmlTextWriter()
        myNoteMarker = MyNoteMarker(osisToHtmlParameters!!, verseInfo!!, htmlTextWriter!!)
        verseHandler = VerseHandler(osisToHtmlParameters!!, verseInfo!!, bookmarkMarkerMock!!, myNoteMarker!!, htmlTextWriter!!)
    }

    /**
     * No attributes. Just start verse, write some content, end verse.
     *
     * <title>The creation</title>
     */
    @Test
    fun testSimpleVerse() {
        osisToHtmlParameters!!.isShowVerseNumbers = true
        val attrs = AttributesImpl()
        attrs.addAttribute(null, null, OSISUtil.OSIS_ATTR_OSISID, null, "Ezek.40.5")
        verseHandler!!.start(attrs)
        htmlTextWriter!!.write("The Creation")
        verseInfo!!.isTextSinceVerse = true
        verseHandler!!.end()
        Assert.assertThat(htmlTextWriter!!.html, CoreMatchers.equalTo(" <span class='verse' id='3.5'><span class='verseNo'>5</span>&#160;<span class='bookmark1'></span><span class='bookmark2'></span>The Creation</span>"))
    }

    /**
     * No attributes. Just start verse, write some content, end verse.
     *
     * <title>The creation</title>
     */
    @Test
    fun testSimpleVerseShowsBookmark() {
        Mockito.`when`(bookmarkMarkerMock!!.bookmarkClasses).thenReturn(Arrays.asList("bookmarkClass"))
        osisToHtmlParameters!!.isShowVerseNumbers = true
        val attrs = AttributesImpl()
        attrs.addAttribute(null, null, OSISUtil.OSIS_ATTR_OSISID, null, "Ezek.40.5")
        verseHandler!!.start(attrs)
        htmlTextWriter!!.write("The Creation")
        verseInfo!!.isTextSinceVerse = true
        verseHandler!!.end()
        Assert.assertThat(htmlTextWriter!!.html, CoreMatchers.equalTo(" <span class='verse bookmarkClass' id='3.5'><span class='verseNo'>5</span>&#160;<span class='bookmark1'></span><span class='bookmark2'></span>The Creation</span>"))
    }

    /**
     * No attributes. Just start verse, write some content, end verse.
     *
     * <title>The creation</title>
     */
    @Test
    fun testNoVerseNumbers() {
        osisToHtmlParameters!!.isShowVerseNumbers = false
        val attrs = AttributesImpl()
        attrs.addAttribute(null, null, OSISUtil.OSIS_ATTR_OSISID, null, "Ezek.40.5")
        verseHandler!!.start(attrs)
        htmlTextWriter!!.write("The Creation")
        verseInfo!!.isTextSinceVerse = true
        verseHandler!!.end()
        Assert.assertThat(htmlTextWriter!!.html, CoreMatchers.equalTo(" <span class='verse' id='3.5'><span class='verseNo position-marker'>&#x200b;</span><span class='bookmark1'></span><span class='bookmark2'></span>The Creation</span>"))
    }

    /**
     * No attributes. Just start verse, write some content, end verse.
     *
     * <title>The creation</title>
     */
    @Test
    fun testNoVerseNumbersButBookmarkStillShown() {
        Mockito.`when`(bookmarkMarkerMock!!.bookmarkClasses).thenReturn(Arrays.asList("bookmarkClass"))
        osisToHtmlParameters!!.isShowVerseNumbers = false
        val attrs = AttributesImpl()
        attrs.addAttribute(null, null, OSISUtil.OSIS_ATTR_OSISID, null, "Ezek.40.5")
        verseHandler!!.start(attrs)
        htmlTextWriter!!.write("The Creation")
        verseInfo!!.isTextSinceVerse = true
        verseHandler!!.end()
        Assert.assertThat(htmlTextWriter!!.html, CoreMatchers.equalTo(" <span class='verse bookmarkClass' id='3.5'><span class='verseNo position-marker'>&#x200b;</span><span class='bookmark1'></span><span class='bookmark2'></span>The Creation</span>"))
    }
}
